File: typescript/standard/04-events-and-stats.ts
/**
 * Standard API Example 04: Events and Statistics
 *
 * Demonstrates Agent event system and statistics tracking.
 * Events allow reactive programming patterns for monitoring.
 *
 * This example demonstrates:
 * - Agent events: job:received, job:completed, job:failed, payment:received
 * - Lifecycle events: starting, started, stopping, stopped, paused, resumed
 * - Statistics: jobsCompleted, totalEarned, averageJobTime, successRate
 * - Real-time monitoring pattern
 *
 * Run: npx ts-node standard/04-events-and-stats.ts
 */

import { Agent, request, serviceDirectory } from '@agirails/sdk';

async function main() {
  console.log('AGIRAILS Standard API - Events and Statistics\n');

  // Create agent
  const agent = new Agent({
    name: 'MonitoredAgent',
    network: 'mock',
  });

  // === LIFECYCLE EVENTS ===
  console.log('Setting up lifecycle event listeners...');

  agent.on('starting', () => {
    console.log('  [EVENT] Agent is starting...');
  });

  agent.on('started', () => {
    console.log('  [EVENT] Agent started successfully!');
  });

  agent.on('paused', () => {
    console.log('  [EVENT] Agent paused');
  });

  agent.on('resumed', () => {
    console.log('  [EVENT] Agent resumed');
  });

  agent.on('stopping', () => {
    console.log('  [EVENT] Agent is stopping...');
  });

  agent.on('stopped', () => {
    console.log('  [EVENT] Agent stopped');
  });

  // === JOB EVENTS ===
  console.log('Setting up job event listeners...');

  agent.on('job:received', (job: any) => {
    console.log(`  [EVENT] Job received: ${job.id.substring(0, 8)}... ($${job.budget})`);
  });

  agent.on('job:completed', (job: any, result: any) => {
    console.log(`  [EVENT] Job completed: ${job.id.substring(0, 8)}...`);
  });

  agent.on('job:failed', (job: any, error: Error) => {
    console.log(`  [EVENT] Job failed: ${job.id.substring(0, 8)}... - ${error.message}`);
  });

  agent.on('job:progress', (jobId: string, percent: number, message?: string) => {
    console.log(`  [EVENT] Job progress: ${jobId.substring(0, 8)}... - ${percent}% ${message || ''}`);
  });

  // === PAYMENT EVENTS ===
  agent.on('payment:received', (amount: number) => {
    console.log(`  [EVENT] Payment received: $${amount.toFixed(2)} USDC`);
  });

  // === ERROR EVENTS ===
  agent.on('error', (error: Error) => {
    console.log(`  [EVENT] Error: ${error.message}`);
  });

  // Register service with progress reporting
  agent.provide('process-data', async (job, ctx) => {
    // Report progress throughout execution
    ctx.progress(0, 'Starting...');

    await new Promise((r) => setTimeout(r, 100));
    ctx.progress(25, 'Loading data...');

    await new Promise((r) => setTimeout(r, 100));
    ctx.progress(50, 'Processing...');

    await new Promise((r) => setTimeout(r, 100));
    ctx.progress(75, 'Finalizing...');

    // Simulate occasional failure
    if (job.input.shouldFail) {
      throw new Error('Simulated processing error');
    }

    ctx.progress(100, 'Complete!');
    return { processed: true, items: job.input.items || 0 };
  });

  // Start agent (triggers lifecycle events)
  console.log('\n--- Starting Agent ---');
  await agent.start();
  serviceDirectory.register('process-data', agent.address);

  // Make some successful requests
  console.log('\n--- Processing Jobs ---\n');

  for (let i = 1; i <= 3; i++) {
    console.log(`Request ${i}:`);
    try {
      await request('process-data', {
        input: { items: i * 10, shouldFail: false },
        budget: 1.0,
      });
    } catch (error) {
      // Handled by event
    }
    console.log('');
  }

  // Make a failing request
  console.log('Request 4 (will fail):');
  try {
    await request('process-data', {
      input: { items: 40, shouldFail: true },
      budget: 1.0,
      timeout: 5000,
    });
  } catch (error) {
    // Handled by event
  }
  console.log('');

  // === STATISTICS ===
  console.log('\n--- Agent Statistics ---\n');

  const stats = agent.stats;
  console.log(`Jobs Received:    ${stats.jobsReceived}`);
  console.log(`Jobs Completed:   ${stats.jobsCompleted}`);
  console.log(`Jobs Failed:      ${stats.jobsFailed}`);
  console.log(`Success Rate:     ${(stats.successRate * 100).toFixed(1)}%`);
  console.log(`Total Earned:     $${stats.totalEarned.toFixed(2)} USDC`);
  console.log(`Avg Job Time:     ${stats.averageJobTime.toFixed(0)}ms`);

  // === BALANCE ===
  console.log('\n--- Agent Balance ---\n');

  const balance = agent.balance;
  console.log(`USDC Balance:     ${balance.usdc}`);
  console.log(`Locked in Escrow: ${balance.locked}`);
  console.log(`Pending:          ${balance.pending}`);

  // Stop agent (triggers lifecycle events)
  console.log('\n--- Stopping Agent ---');
  serviceDirectory.unregister('process-data', agent.address);
  await agent.stop();

  console.log('\nEvents and stats demo complete!');
}

main().catch(console.error);

File: python/standard/04_events_and_stats.py
#!/usr/bin/env python3
"""
Standard API Example 04: Events and Stats

Demonstrates the Agent event system and real-time statistics tracking.

This example demonstrates:
- Event listener registration
- Lifecycle events (started, stopped, paused, resumed)
- Job events (received, completed, failed, progress)
- Payment events
- Real-time statistics

Run: python standard/04_events_and_stats.py
"""

import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.utils.helpers import clear_mock_state


async def main() -> None:
    print("AGIRAILS Standard API - Events and Stats\n")

    clear_mock_state()

    from agirails.level1 import Agent, AgentConfig, Job, JobContext

    # Create agent
    agent = Agent(
        AgentConfig(
            name="EventsDemo",
            description="Demonstrates event handling",
            network="mock",
        )
    )

    # Event counters for demo
    event_counts = {
        "started": 0,
        "stopped": 0,
        "paused": 0,
        "resumed": 0,
        "job:received": 0,
        "job:completed": 0,
        "job:failed": 0,
        "payment:received": 0,
        "error": 0,
    }

    # 1. Register lifecycle event handlers
    def on_started():
        event_counts["started"] += 1
        print("[EVENT] Agent started")

    def on_stopped():
        event_counts["stopped"] += 1
        print("[EVENT] Agent stopped")

    def on_paused():
        event_counts["paused"] += 1
        print("[EVENT] Agent paused")

    def on_resumed():
        event_counts["resumed"] += 1
        print("[EVENT] Agent resumed")

    agent.on("started", on_started)
    agent.on("stopped", on_stopped)
    agent.on("paused", on_paused)
    agent.on("resumed", on_resumed)

    # 2. Register job event handlers
    def on_job_received(job: Job):
        event_counts["job:received"] += 1
        print(f"[EVENT] Job received: {job.id[:12]}... (service: {job.service})")

    def on_job_completed(job: Job, result: dict):
        event_counts["job:completed"] += 1
        print(f"[EVENT] Job completed: {job.id[:12]}...")

    def on_job_failed(job: Job, error: Exception):
        event_counts["job:failed"] += 1
        print(f"[EVENT] Job failed: {job.id[:12]}... - {error}")

    def on_job_progress(job: Job, progress: int, message: str):
        print(f"[EVENT] Job progress: {job.id[:12]}... - {progress}% - {message}")

    agent.on("job:received", on_job_received)
    agent.on("job:completed", on_job_completed)
    agent.on("job:failed", on_job_failed)
    agent.on("job:progress", on_job_progress)

    # 3. Register payment event handler
    def on_payment(amount: float, tx_id: str):
        event_counts["payment:received"] += 1
        print(f"[EVENT] Payment received: ${amount:.2f} (tx: {tx_id[:12]}...)")

    agent.on("payment:received", on_payment)

    # 4. Register error handler
    def on_error(error: Exception):
        event_counts["error"] += 1
        print(f"[EVENT] Error: {error}")

    agent.on("error", on_error)

    # 5. Register services
    @agent.provide("fast-service")
    async def handle_fast(job: Job, ctx: JobContext) -> dict:
        ctx.progress(50, "Processing...")
        await asyncio.sleep(0.1)
        ctx.progress(100, "Done")
        return {"result": "fast", "duration_ms": 100}

    @agent.provide("slow-service")
    async def handle_slow(job: Job, ctx: JobContext) -> dict:
        for i in range(5):
            ctx.progress(i * 20, f"Step {i + 1}/5")
            await asyncio.sleep(0.2)
        ctx.progress(100, "Complete")
        return {"result": "slow", "duration_ms": 1000}

    print("Registered event handlers for:")
    print("  - Lifecycle: started, stopped, paused, resumed")
    print("  - Jobs: received, completed, failed, progress")
    print("  - Payments: received")
    print("  - Errors")
    print()

    # 6. Start agent (triggers 'started' event)
    print("Starting agent...")
    await agent.start()
    print()

    # 7. Demonstrate lifecycle events
    print("Testing lifecycle events...")
    await asyncio.sleep(0.5)

    agent.pause()  # Triggers 'paused' event
    await asyncio.sleep(0.5)

    agent.resume()  # Triggers 'resumed' event
    await asyncio.sleep(0.5)
    print()

    # 8. Show real-time stats
    print("=" * 50)
    print("Real-time Agent Statistics")
    print("=" * 50)
    print()

    def print_stats():
        stats = agent.stats
        print(f"Jobs Received:    {stats.jobs_received}")
        print(f"Jobs Completed:   {stats.jobs_completed}")
        print(f"Jobs Failed:      {stats.jobs_failed}")
        print(f"Jobs In Progress: {stats.jobs_in_progress}")
        print(f"Total Earned:     ${stats.total_earned:.2f}")
        print(f"Total Spent:      ${stats.total_spent:.2f}")
        print(f"Success Rate:     {stats.success_rate:.1f}%")
        print(f"Avg Job Time:     {stats.average_job_time}ms")

    print("Initial stats:")
    print_stats()
    print()

    # 9. Simulate some activity
    print("Simulating agent activity...")
    await asyncio.sleep(2)
    print()

    print("Stats after activity:")
    print_stats()
    print()

    # 10. Event summary
    print("=" * 50)
    print("Event Summary")
    print("=" * 50)
    for event_name, count in event_counts.items():
        print(f"  {event_name}: {count}")
    print()

    # 11. Stop agent (triggers 'stopped' event)
    print("Stopping agent...")
    await agent.stop()

    print("\nEvents demo complete!")


if __name__ == "__main__":
    asyncio.run(main())
